tests/framework/prototype.py

```diff
@@ -0,0 +1,279 @@
+
+def search_path(cr, steps):
+    curr = cr
+    for step in steps:
+        if step.isnumeric():
+            int_step = int(step)
+            if int_step >= len(curr) or int_step < 0:
+                return None
+            curr = curr[int(step)]
+        elif step in curr:
+            curr = curr[step]
+        else:
+            return None
+    return curr
+
+'''
+Functions for cluster preparation. Typical workflow:
+  Delete KinD cluster -> Create KinD cluster -> Install CRD -> Download Images (from DockerHub) ->
+  Load images into KinD cluster -> Install KubeRay operator
+'''
+def delete_kind_cluster() -> None:
+    """Delete a KinD cluster"""
+    os.system("kind delete cluster")
+
+def create_kind_cluster():
+    """Create a KinD cluster"""
+    os.system("kind create cluster")
+    os.system("kubectl wait --for=condition=ready pod -n kube-system --all --timeout=900s")
+
+def install_crd():
+    """Install Custom Resource Definition (CRD)"""
+    KUBERAY_VERSION = "v0.3.0"
+    os.system("kubectl create -k \"github.com/ray-project/kuberay/manifests/cluster-scope-resources?" +
+              "ref={}&timeout=90s\"".format(KUBERAY_VERSION))
+
+def download_images(images):
+    """Download Docker images from DockerHub"""
+    docker_client = docker.from_env()
+    for image in images:
+        docker_client.images.pull(image)
+    docker_client.close()
+
+def kind_load_images(images):
+    """Load downloaded images into KinD cluster"""
+    for image in images:
+        os.system(f'kind load docker-image {image}')
+
+def install_kuberay_operator():
+    KUBERAY_VERSION = "v0.3.0"
+    os.system("kubectl apply -k \"github.com/ray-project/kuberay/manifests/base?" +
+              "ref={}&timeout=90s\"".format(KUBERAY_VERSION))
+
+def check_cluster_exist():
+    return os.system("kubectl cluster-info --context kind-kind") == 0 
+
+'''
+Configuration Test Framework Abstractions: (1) Mutator (2) Rule (3) RuleSet (4) CREvent
+'''
+
+# Mutator: Mutator will start to mutate from `baseCR`. `patch_list` is a list of JsonPatch, and you can
+#          specify multiple fields that want to mutate in a single JsonPatch.
+class Mutator:
+    def __init__(self, baseCR, patch_list: list[jsonpatch.JsonPatch]):
+        self.baseCR = baseCR
+        self.patch_list = patch_list
+    # Generate a new cr by applying the json patch to `cr`. 
+    def mutate(self):
+        for patch in self.patch_list:
+            yield patch.apply(self.baseCR)
+
+# Rule: Rule is used to check whether the actual cluster state is the same as our expectation after a CREvent.
+#       We can infer the expected state by CR YAML file, and get the actual cluster state by Kubernetes API.
+# Example: "HeadPodNameRule"
+class Rule:
+    def __init__(self):
+        pass
+    # The rule will only be checked when `trigger_condition` is true. For example, we will only check
+    # "HeadPodNameRule" when "spec.headGroupSpec" is defined in CR YAML file.
```

Comment:
spec.headGroupSpec is actually always defined (it's are required field).

Could you explain more about why trigger_condition is a useful concept? (Could you just write the equivalent logic within the rule assertion?)

Reply:
Explanation: It is a common enough pattern that certain functionality only needs to be tested for configs satisfying certain condition.
This method is fine to keep.

Reply:
Added a section "Why do we need trigger_condition? Could we write equivalent logic within assertRule?" in the design doc.